OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java

```diff
@@ -572,139 +443,177 @@ private static OSEmailSubscriptionState getLastEmailSubscriptionState(Context co
+      if (appId == null) {
+         OneSignal.onesignalLog(LOG_LEVEL.WARN, "appContext set, but please call setAppId(appId) with a valid appId to complete OneSignal init!");
+         return;
+      }
+
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppContext(context) successful and appId is set, continuing OneSignal init...");
+      init();
    }
 
-   public static void init(Context context, String googleProjectNumber, String oneSignalAppId) {
-      init(context, googleProjectNumber, oneSignalAppId, null, null);
+   public static void setNotificationWillShowInForegroundHandler(NotificationWillShowInForegroundHandler callback) {
+      notificationWillShowInForegroundHandler = callback;
    }
 
-   public static void init(Context context, String googleProjectNumber, String oneSignalAppId, NotificationOpenedHandler notificationOpenedHandler) {
-      init(context, googleProjectNumber, oneSignalAppId, notificationOpenedHandler, null);
+   public static void setNotificationOpenedHandler(NotificationOpenedHandler callback) {
+      notificationOpenedHandler = callback;
+      if (appContext != null)
+        setAppContext(appContext);
    }
 
-   public static void init(Context context, String googleProjectNumber, String oneSignalAppId, NotificationOpenedHandler notificationOpenedHandler, NotificationReceivedHandler notificationReceivedHandler) {
-      mInitBuilder = createInitBuilder(notificationOpenedHandler, notificationReceivedHandler);
-      OneSignal.setAppContext(context);
-      setupPrivacyConsent(context);
+   public static void setInAppMessageClickHandler(InAppMessageClickHandler callback) {
+      inAppMessageClickHandler = callback;
+   }
 
+   /**
+    * Called after setAppId and setAppContext, depending on which one is called last (order does not matter)
+    */
+   synchronized private static void init() {
       if (requiresUserPrivacyConsent()) {
-         OneSignal.Log(LOG_LEVEL.VERBOSE, "OneSignal SDK initialization delayed, user privacy consent is set to required for this application.");
-         delayedInitParams = new DelayedConsentInitializationParameters(context, googleProjectNumber, oneSignalAppId, notificationOpenedHandler, notificationReceivedHandler);
+         OneSignal.Log(LOG_LEVEL.WARN, "OneSignal SDK initialization delayed, user privacy consent is set to required for this application.");
+         delayedInitParams = new DelayedConsentInitializationParameters(appContext, googleProjectNumber, appId, notificationWillShowInForegroundHandler, notificationOpenedHandler);
+         // Set app id null since OneSignal was not init fully
+         appId = null;
          return;
       }
 
-      mInitBuilder = createInitBuilder(notificationOpenedHandler, notificationReceivedHandler);
-
       if (!isGoogleProjectNumberRemote())
-         mGoogleProjectNumber = googleProjectNumber;
+         googleProjectNumber = "";
```

Comment:
Lets remove these 2 lines, we don't need to set googleProjectNumber to a blank string

Reply:
Okay sounds good